### PR TITLE
Expose pre-state Version via TransactEx

### PR DESCRIPTION
**Pull request by bartelink**
At present, the `Equinox.Decider` interface only exposes the outgoing `Version`, which means the only way to determine the incoming Version is by indirect means such as:
- checking the `Index` of the freshest event (can be incorrect if an Event that's not in the DU is encountered)
- inferring it based on the post-`Version` passed to the `mapResult` (won't work well with transmute operations or rolling snapshots)

This proposal provides the incoming extended state via a new `TransactEx` API, enabling one to process the value directly.

As this is only for V3, it makes sense to remove `TransactAsyncEx` as both APIs are likely extremely rare, and having 2 mysteriously named functions with complex signatures is rarely better than one.

cc @dharmaturtle

### CODE REVIEW

File: src/Equinox/Equinox.fs [hunk 1]
```diff
@@ -10,42 +10,50 @@ type MaxResyncsExhaustedException(count) =
 /// Central Application-facing API. Wraps the handling of decision or query flows in a manner that is store agnostic
 type Stream<'event, 'state>
     (   log, stream : IStream<'event, 'state>, maxAttempts : int,
-        [<Optional; DefaultParameterValue(null)>] ?createAttemptsExhaustedException,
+        [<Optional; DefaultParameterValue(null)>] ?createAttemptsExhaustedException : int -> exn,
```
Comment (bartelink) [inline src/Equinox/Equinox.fs:13]: this signature fix should be a separate PR, and probably should only happen in V3
(the restriction for the exception to inherit from `MaxResyncsExhaustedException` was unintentional)

**Comment by bartelink** [comment]
This PR is an example to illustrate an approach in a discussion on the DDD-CQRS-ES Slack - my instinct is not to add this until we validate a scenario for which this truly is the best solution.

**Comment by bartelink** [comment]
@dharmaturtle I'll likely take the cleanup aspects of this PR, and the FAQ from this and merge it after #272 for inclusion in the V3 RC

As noted previously, any concept that gets exposed in the public API has a cost in terms of adding stuff for readers of the code to traverse. That doesn't always make it a bad thing - sometimes naming or mentioning the concept can help people's thinking even if they opt not to actually use the facility in the end.

Please let me know if you've found a use case for needing to know the pre-version or have any other thoughts on this.

**Comment by dharmaturtle** [comment]
I've thrown pre-release binaries into source control before. I don't think it's a bad practice, and [neither does Mark Seemann](https://blog.ploeh.dk/2014/01/29/nuget-package-restore-considered-harmful/). There's no particular rush :)

**Comment by bartelink** [comment]
Thats an interesting idea; in this instance it's more a concern of whether the feature should be added in the first instance than the release vehicle itself though.

If you wanted/needed to trial it, PR builds yield [alpha-badged] artifacts and can be released, though we try not to do that too often in order to avoid stressing nuget's performance.

On re-reading, I'm beginning to like the OP suggestion of exposing it as a breaking change in [`TransactAsyncEx`](https://github.com/jet/equinox/blob/master/src/Equinox/Equinox.fs#L47) - that provides the escape hatch just in case, without suggesting that it's part of any mainline API.

File: src/Equinox/Equinox.fs (hunk 1 above)
Comment (bartelink) [reply src/Equinox/Equinox.fs]: opted to leave it here
